coderd/schedule/autostop.go

```diff
@@ -0,0 +1,329 @@
+	// amount of hours is impacted by daylight savings. Even though daylight
+	// savings is usually only an hour difference, this calculation is used to
+	// get the current week number and could result in an entire week getting
+	// skipped if the calculation is off by an hour.
+	//
+	// Old naive algorithm: weeksSinceEpoch := int64(since.Hours() / (24 * 7))
+
+	// Get days since epoch. Start with a negative number of days, as we want to
+	// subtract the YearDay() of the epoch itself.
+	days := -epoch.YearDay()
+	for i := epoch.Year(); i < now.Year(); i++ {
+		startOfNextYear := time.Date(i+1, 1, 1, 0, 0, 0, 0, now.Location())
+		if startOfNextYear.Year() != i+1 {
+			return 0, xerrors.New("overflow calculating weeks since epoch")
+		}
+		endOfThisYear := startOfNextYear.AddDate(0, 0, -1)
+		if endOfThisYear.Year() != i {
+			return 0, xerrors.New("overflow calculating weeks since epoch")
+		}
+
+		days += endOfThisYear.YearDay()
+	}
+	// Add this year's days.
+	days += now.YearDay()
+
+	// Ensure that the number of days is positive.
+	if days < 0 {
+		return 0, xerrors.New("overflow calculating weeks since epoch")
+	}
+
+	// Divide by 7 to get the number of weeks.
+	weeksSinceEpoch := int64(days / 7)
+	return weeksSinceEpoch, nil
+}
+
+// GetMondayOfWeek gets the Monday (0:00) of the n-th week since epoch.
+func GetMondayOfWeek(loc *time.Location, n int64) (time.Time, error) {
+	if n < 0 {
+		return time.Time{}, xerrors.New("weeks since epoch must be positive")
+	}
+	epoch := TemplateRestartRequirementEpoch(loc)
+	monday := epoch.AddDate(0, 0, int(n*7))
+
+	y, m, d := monday.Date()
+	monday = time.Date(y, m, d, 0, 0, 0, 0, loc)
+	if monday.Weekday() != time.Monday {
+		return time.Time{}, xerrors.Errorf("calculated incorrect Monday for week %v since epoch (actual weekday %q)", n, monday.Weekday())
+	}
+	return monday, nil
+}
+
+// GetNextApplicableMondayOfNWeeks gets the next Monday (0:00) of the next week
+// divisible by n since epoch. If the next applicable week is invalid for any
+// reason, the week after will be used instead (up to 2 attempts).
+//
+// If the current week is divisible by n, then the provided time is returned as
+// is.
+//
+// The timezone embedded in the time object is used to determine the epoch.
+func GetNextApplicableMondayOfNWeeks(now time.Time, n int64) (time.Time, error) {
+	// Get the current week number.
+	weeksSinceEpoch, err := WeeksSinceEpoch(now)
+	if err != nil {
+		return time.Time{}, xerrors.Errorf("get current week number: %w", err)
+	}
+
+	// Get the next week divisible by n.
+	remainder := weeksSinceEpoch % n
+	week := weeksSinceEpoch + (n - remainder)
+	if remainder == 0 {
+		return now, nil
+	}
+
+	// Loop until we find a week that doesn't fail.
+	var lastErr error
+	for i := int64(0); i < 3; i++ {
+		monday, err := GetMondayOfWeek(now.Location(), week+i)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+
+		return monday, nil
+	}
```

Comment:
Do you have any example of why this would fail?

Reply:
This is my only open question.

Reply:
To be honest, I'm not sure how this could fail, but the `GetMondayOfWeek` algorithm just adds `7 * n` days to the epoch which seems kinda naive to me. This is tested though for the next 100 years in all of those timezones so I think it's stable, I'll add a comment saying this is probably not needed but here just in case.

Reply:
Perfect. Was just making sure I wasn't missing something.